olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java

```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+                        valueMatch = Util.doubleEquals(this.innerDoubleValue.get(), that.innerDoubleValue.get());
+                    } else if (this.innerDateTime.isPresent() && that.innerDateTime.isPresent()) {
+                        valueMatch = this.innerDateTime.get().equals(that.innerDateTime.get());
+                    } else if (this.innerTime.isPresent() && that.innerTime.isPresent()) {
+                        valueMatch = this.innerTime.get().equals(that.innerTime.get());
+                    } else {
+                        valueMatch = this.innerValue.equals(that.innerValue);
+                    }
+                    if(!valueMatch) {
+                        logger.fine(String.format("Property Value mismatch: %s and %s", this.reportString(), that.reportString()));
+                    }
+                    return valueMatch;
+                } else {
+                    logger.fine(String.format("Reference/Value mismatch: %s and %s", this.reportString(), that.reportString()));
+                    return false;
+                }
+            } else {
+                logger.fine("b is not a DerefedProperty b.class: " + o.getClass());
+                return false;
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "DerefedProperty(" +
+                    "isDerefed=" + isDerefed +
+                    ", innerConf=" + innerConf +
+                    ", innerValue='" + innerValue + '\'' +
+                    ", propName='" + propName + '\'' +
+                    ", locationContext='" + locationContext + '\'' +
+                    ')';
+        }
+    }
+
+    /**
+     * See {@link #structuralEquals(List, List, String, String)} for description of behavior.
+     */
+    private static boolean innerStructuralEquals(Map<String, ConfigurationData> a, Map<String, ConfigurationData> b, String aName, String bName) {
+        Optional<StructuralConfigurationData> aRootOpt = StructuralConfigurationData.fromName(a, aName);
+        Optional<StructuralConfigurationData> bRootOpt = StructuralConfigurationData.fromName(b, bName);
+        if (aRootOpt.isPresent() && bRootOpt.isPresent()) {
+            return aRootOpt.get().equals(bRootOpt.get());
+        } else {
+            if (aRootOpt.isPresent()) {
+                logger.fine(String.format("Component with name: %s not found in b", bName));
+            } else {
+                logger.fine(String.format("Component with name: %s not found in a", aName));
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Checks whether two ConfigurationData objects are 'structurally equal'. Two objects are structurally
+     * equal when their classNames are the same and when all of their properties are equal or, if those
+     * properties refer to another ConfigurationData by name, if all of their properties are structurally
+     * equal recursively.
```

Comment:
Add a note saying it does not compare the serializedForm, importable, exportable, leaseTime or entries attributes of the configuration data. Everything apart from serializedForm is going to go away in the next version anyway, but it's worth noting here.